modelmapper/signature.py

```diff
@@ -0,0 +1,72 @@
+from collections import Mapping
+from decimal import Decimal
+
+import mmh3
+
+
+def generate_row_signature(row, model=None, ignore_fields=None):
+    """ Generates a 64 bit hash of the given row
+        Arguments:
+            row: A dictionary or list of tuples
+            model: Sqlalchemy model that coresponses to the given row
+            ignore_fields: Fields to be ignored in the calcuation of the hash
+        Returns:
+            Integer: the hash value of the given row
+    """
+    if isinstance(row, list):
+        row_dict = dict(row)
+    elif isinstance(row, Mapping):
+        row_dict = row
+    else:
+        raise TypeError('Row needs to be a list of tuples or a dictionary')
+
+    default_dropped_row = drop_model_defaults(row_dict, model)
+    normalized_row = normalize_decimal_columns(default_dropped_row)
+    sorted_row = sort_row_values(normalized_row)
+    row_bytes = get_byte_str_of_row(sorted_row, ignore_fields)
+    return get_hash_of_bytes(row_bytes)
+
+
+def normalize_decimal_columns(row):
+    """Remove trailing zeros from Decimal fields"""
+    normalized_row = {}
+    for k, v in row.items():
+        if isinstance(v, Decimal):
+            normalized_row[k] = v.normalize()
+        else:
+            normalized_row[k] = v
+    return normalized_row
+
+
+def drop_model_defaults(row, model):
+    """Remove all columns from the row that are equal to their sqlalchemy default"""
+    if model is None:
+        new_row = row.copy()
```

Comment:
Why do we need to copy the row?

Reply:
did not want to edit the original dict being provided so made a new one